test/parallel/test-zlib-flush-drain.js

```diff
@@ -0,0 +1,49 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const zlib = require('zlib');
+const path = require('path');
+
+const bigData = new Buffer(10240).fill('x');
+
+const opts = {
+  level: 0,
+  highWaterMark: 16
+};
+
+const deflater = zlib.createDeflate(opts);
+
+// shim deflater.flush so we can count times executed
+var flushCount = 0;
+
+const flush = deflater.flush;
+deflater.flush = function(kind, callback) {
+  flushCount++;
+  flush.call(this, kind, callback);
+};
+
+deflater.write(bigData);
+
+const ws = deflater._writableState;
+const beforeFlush = ws.needDrain;
+var afterFlush = ws.needDrain;
+var drained = false;
+
+deflater.flush(function(err) {
+  afterFlush = ws.needDrain;
+});
+
+deflater.on('drain', function() {
+  drained = true;
```

Comment:
drain should only execute once if I recall correctly, but I can double check that